src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs

```diff
@@ -267,23 +272,51 @@ private void EnsureServer()
             }
         }
 
-        public void Dispose()
+        public async Task StopAsync(CancellationToken cancellationToken)
         {
+            if (_stopped)
+            {
+                return;
+            }
+            _stopped = true;
+
             _logger?.Shutdown();
 
+            if (!cancellationToken.CanBeCanceled)
+            {
+                cancellationToken = new CancellationTokenSource(Options.ShutdownTimeout).Token;
+            }
+
             // Fire IApplicationLifetime.Stopping
             _applicationLifetime?.StopApplication();
 
+            try
+            {
+                await Server?.StopAsync(cancellationToken);
+            }
+            catch (Exception ex)
+            {
+                _logger?.ServerShutdownException(ex);
```

Comment:
Do we think that WebHost.Dispose() should catch any errors from IServer.Stop without rethrowing? If IServer.StopStop is called first, then we don't have to worry about WebHost.Dispose() throwing anyway.